include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -67,36 +41,94 @@ struct exchangeValues<SharedMemoryAdaptor, float32_t>
         {
             hi.real(exchanged.x);
             hi.imag(exchanged.y);
-        }      
-    }
-};
-
-template<typename SharedMemoryAdaptor>
-struct exchangeValues<SharedMemoryAdaptor, float64_t>
-{
-    static void __call(NBL_REF_ARG(complex_t<float64_t>) lo, NBL_REF_ARG(complex_t<float64_t>) hi, uint32_t threadID, uint32_t stride, NBL_REF_ARG(SharedMemoryAdaptor) sharedmemAdaptor)
-    {
-        const bool topHalf = bool(threadID & stride);
-        // pack into `float64_t2` because ternary operator doesn't support structs
-        float64_t2 exchanged = topHalf ? float64_t2(lo.real(), lo.imag()) : float64_t2(hi.real(), hi.imag());                    
-        shuffleXor<SharedMemoryAdaptor, float64_t2 >::__call(exchanged, stride, sharedmemAdaptor);
-        if (topHalf)
-        {
-            lo.real(exchanged.x);
-            lo.imag(exchanged.y);
         }
-        else
-        {
-            hi.real(exchanged.x);
-            hi.imag(exchanged.y);
-        }      
     }
 };
 
 // Get the required size (in number of uint32_t elements) of the workgroup shared memory array needed for the FFT
 template <typename scalar_t, uint32_t WorkgroupSize>
 NBL_CONSTEXPR uint32_t SharedMemoryDWORDs = (sizeof(complex_t<scalar_t>) / sizeof(uint32_t))  * WorkgroupSize;
 
+
+template<uint32_t N, uint32_t H>
+enable_if_t<H <= N, uint32_t> bitShiftRightHigher(uint32_t i)
+{
+    // Highest H bits are numbered N-1 through N - H
+    // N - H is then the middle bit
+    // Lowest bits numbered from 0 through N - H - 1
+    uint32_t low = i & ((1 << (N - H)) - 1);
+    uint32_t mid = i & (1 << (N - H));
+    uint32_t high = i & ~((1 << (N - H + 1)) - 1);
```

Comment:
precompute some of these, like `midMask = 1<<(N-H)`, also `~((1 << (N - H + 1))-1)` is just `(2<<N)-midMask` if you don't allow for `N==32`

Reply:
Aren't they precomputed at compile time?

Reply:
they do get precomputed, its not about perf, its about readibility